lib/Auth/Process/DynamicAssurance.php

```diff
@@ -141,67 +152,69 @@ public function process(&$state)
         assert('is_array($state)');
         assert('array_key_exists("Attributes", $state)');
 
-        // Return early if assurance matches one of the well known values
-        // Candidates
-        if (
-            !empty($state['Attributes'][$this->attribute])
-            && !empty(array_intersect($state['Attributes'][$this->attribute], $this->candidates))
-        ) {
-            SimpleSAML_Logger::debug(
-                "[DynamicAssurance] Assurance matches known value: "
-                . var_export(array_intersect($state['Attributes'][$this->attribute], $this->candidates), true)
-            );
-            return;
+        // Append in the Assurance Attribute all the configured values
+        $assurance_from_candidates = array();
+        foreach ($this->attributeMap as $attribute => $oidCandidates) {
+            if(is_array($oidCandidates)) {                  // Assurance Mappings
+                if (!empty($state['Attributes'][$attribute])) {
+                    foreach ($state['Attributes'][$attribute] as $assurance_oid) {
+                        if (!empty($oidCandidates[$assurance_oid])) {
+                            $assurance_from_candidates = array_merge(
+                                $assurance_from_candidates,
+                                $oidCandidates[$assurance_oid]
+                            );
+                        }
+                    }
+                }
+            } elseif(is_string($oidCandidates)) {           // Pass Through
+                $passthrough_values = preg_grep($oidCandidates, $state['Attributes'][$attribute]);
+                $assurance_from_candidates = array_merge(
+                    $assurance_from_candidates,
+                    $passthrough_values
+                );
+            }
         }
 
-        $assurance = $this->defaultAssurance;
-
-        // Elevate assurance?
         // If the module is active on a bridge,
         // $state['saml:sp:IdP'] will contain an entry id for the remote IdP.
         if (!empty($state['saml:sp:IdP'])) {
             $idpEntityId = $state['saml:sp:IdP'];
-            $idpMetadata = SimpleSAML_Metadata_MetaDataStorageHandler::getMetadataHandler()->getMetaData($idpEntityId, 'saml20-idp-remote');
+            $idpMetadata = SimpleSAML_Metadata_MetaDataStorageHandler::getMetadataHandler()->getMetaData(
+                $idpEntityId,
+                'saml20-idp-remote'
+            );
         } else {
             $idpEntityId = $state['Source']['entityid'];
```

Comment:
Retrieve the IdP tags *only* if the idpTagMap is non-empty

Reply:
Done